### ISSUE TITLE
temporal embedding design in VideoVAE

### ISSUE THREAD

**Issue by dc3ea9f** [PKU-YuanGroup/Open-Sora-Plan#246]
@qqingzheng 
Hello, thank you for your valuable contributions! I have some inquiries regarding the temporal embedding designs employed in VideoVAE.

In the provided model from [huggingface](https://huggingface.co/LanguageBind/Open-Sora-Plan-v1.0.0/blob/main/vae/config.json), the upsampling/downsampling layers are implemented as `TimeDownsample2x`/`TimeUpsample2x`. However, it appears that these layers lack parameters, which might result in an insufficient recovery of precise motion within the VAE framework, as discussed in https://github.com/PKU-YuanGroup/Open-Sora-Plan/pull/145#issuecomment-2032301161. I suspect that the absence of parameters for temporal modeling could potentially lead to motion blur in VAE reconstructions.

Upon further investigation, I noticed alternative implementations in [.../modules/updownsample.py](https://github.com/PKU-YuanGroup/Open-Sora-Plan/blob/main/opensora/models/ae/videobase/modules/updownsample.py), namely `TimeDownsampleRes2x` and `TimeUpsampleRes2x`. These implementations seem promising in mitigating motion blur during reconstruction. However, I have two specific questions regarding these implementations:
1. **Regarding TimeDownsampleRes2x:** Why is `Conv3D` utilized instead of `CausalConv3D`, as seen in `TimeUpsampleRes2x`? It seems logical for these two layers to employ consistent architectures.
2. **Regarding TimeUpsampleRes2x:** Why is `CausalConv3D` employed only for frames $[1, N)$? This design choice implies that information from frame 0 cannot be fully propagated to successive frames. Does this approach provide sufficient temporal modeling, or might it be lacking in capturing the complete temporal dynamics?

I appreciate your insights on these matters. Thank you again for your efforts in advancing this project!

**Comment by qqingzheng** [comment]
Thank you for your attention and in-depth understanding of this project.

# Q1

For your first question, actually, this part also utilizes CausalConv3D, as you can see from the code snippet below. Before performing the convolution, a replication padding of `k-1` is also conducted. Therefore, using CausalConv3D and Conv3D here is equivalent. The reason for not directly using CausalConv3D is because this module requires an Avgpool, which also has the replication padding process.

```python
class TimeDownsampleRes2x(nn.Module):
    def __init__(
        self,
        in_channels,
        out_channels,
        kernel_size: int = 3,
        mix_factor: float = 2,
    ):
        super().__init__()
        self.kernel_size = cast_tuple(kernel_size, 3)
        self.avg_pool = nn.AvgPool3d((kernel_size,1,1), stride=(2,1,1))
        self.conv = nn.Conv3d(
            in_channels, out_channels, self.kernel_size, stride=(2,1,1), padding=(0,1,1)
        )
        self.mix_factor = torch.nn.Parameter(torch.Tensor([mix_factor]))
    
    def forward(self, x):
        alpha = torch.sigmoid(self.mix_factor)
        first_frame_pad = x[:, :, :1, :, :].repeat(
            (1, 1, self.kernel_size[0] - 1, 1, 1)
        )
        x = torch.concatenate((first_frame_pad, x), dim=2)
        return alpha * self.avg_pool(x) + (1 - alpha) * self.conv(x)
```

# Q2

I'm not sure if I misunderstood your question, but the convolution here includes frame 0. Here, interpolation upsampling is not performed on the first frame. Subsequent frames are interpolation upsampled before being concatenated with the first frame, and then input into CausalConv3D. 

```python
class TimeUpsampleRes2x(nn.Module):
    ...
    def forward(self, x):
        alpha = torch.sigmoid(self.mix_factor)
        if x.size(2) > 1:
            x,x_= x[:,:,:1],x[:,:,1:]
            x_= F.interpolate(x_, scale_factor=(2,1,1), mode='trilinear')
            x = torch.concat([x, x_], dim=2)
        return alpha * x + (1-alpha) * self.conv(x)
```

**Comment by qqingzheng** [comment]
It appears that using learnable temporal upsampling layers can alleviate motion blur. Other methods that can help include the design of the discriminator and the adjustment of the mix factor.

Currently, DepthToSpace/Temporal layers has not been attempted. If you conduct experiments in these methods in the future, welcome to contribute. 

We are currently still working on experiments, and we welcome you to offer deeper insights and share your experimental experiences. 😊

**Comment by qqingzheng** [comment]
At present, experiments on the mix factor are still ongoing and there are no stable conclusions yet. They will be public in the next version. You can check # 168 for initialization of Image VAE.
